src/meta/src/barrier/worker.rs

```diff
@@ -311,19 +313,83 @@ impl<C: GlobalBarrierWorkerContext> GlobalBarrierWorker<C> {
+                                } else {
+                                    info!(database_id = database_id.database_id, "database removed after reloading empty runtime info");
+                                    entering_initializing.remove();
+                                }
+                            }
+                            CheckpointControlEvent::EnteringRunning(entering_running) => {
+                                self.context.mark_ready(Some(entering_running.database_id()));
+                                entering_running.enter();
+                            }
+                        }
+                    };
+                    if let Err(e) = result {
+                        self.failure_recovery(e).await;
+                    }
+                }
+                (worker_id, resp_result) = self.control_stream_manager.next_response() => {
+                     let resp = match resp_result {
+                        Err(e) => {
                             if self.checkpoint_control.is_failed_at_worker_err(worker_id) {
                                 let errors = self.control_stream_manager.collect_errors(worker_id, e).await;
                                 let err = merge_node_rpc_errors("get error from control stream", errors);
-                                self.report_collect_failure(&err);
+                                Self::report_collect_failure(&self.env, &err);
                                 self.failure_recovery(err).await;
                             }  else {
                                 warn!(worker_id, "no barrier to collect from worker, ignore err");
                             }
+                            continue;
+                        }
+                        Ok(resp) => resp,
+                    };
+                    let result: MetaResult<()> = try {
+                        match resp {
+                            Response::CompleteBarrier(resp) => {
+                                self.checkpoint_control.barrier_collected(resp, &mut self.control_stream_manager)?;
+                            },
+                            Response::ReportDatabaseFailure(resp) => {
+                                if !self.enable_recovery {
+                                    panic!("database failure reported but recovery not enabled: {:?}", resp)
+                                }
+                                if let Some(entering_recovery) = self.checkpoint_control.on_report_failure(resp, &mut self.control_stream_manager) {
+                                    let database_id = entering_recovery.database_id();
+                                    warn!(database_id = database_id.database_id, "database entering recovery");
+                                    self.context.abort_and_mark_blocked(Some(database_id), RecoveryReason::Failover(anyhow!("reset database: {}", database_id).into()));
```

Comment:
cc @yezizp2012 Will it be an issue if we remove the [barrier_manager.check_status_running](https://github.com/risingwavelabs/risingwave/blob/7baa0d19cc4b40e0a69c0fc6551e0af8f6d4b68c/src/meta/src/barrier/manager.rs#L83) check completely? Previously the check seems to be used to avoid [DDL](https://github.com/risingwavelabs/risingwave/blob/7baa0d19cc4b40e0a69c0fc6551e0af8f6d4b68c/src/meta/src/rpc/ddl_controller.rs#L293) and scale ([1](https://github.com/risingwavelabs/risingwave/blob/7baa0d19cc4b40e0a69c0fc6551e0af8f6d4b68c/src/meta/service/src/scale_service.rs#L112), [2](https://github.com/risingwavelabs/risingwave/blob/7baa0d19cc4b40e0a69c0fc6551e0af8f6d4b68c/src/meta/src/rpc/ddl_controller.rs#L395)). However, the check is not 100% accurate because it is possible that recovery is triggered immediately after the check has passed. 

In this case, we still need to rely on the [validation in barrier manager](https://github.com/risingwavelabs/risingwave/blob/7baa0d19cc4b40e0a69c0fc6551e0af8f6d4b68c/src/meta/src/barrier/schedule.rs#L101) to make sure the corresponding command is not allowed. In other words, if we change to only rely on the barrier manager validation, will it be a problem? We may need to check whether there can be partial metadata written before barrier injection and whether these partial metadata will be cleaned up correctly on barrier injection/collection failure.